lace/test_geometry.py

```diff
@@ -187,3 +187,24 @@ def test_predict_body_units_cm(self):
         mesh = Mesh()
         mesh.v = np.array([[150, 100, 50], [-150, 50, 150]])
         self.assertEqual(mesh.predict_body_units(), 'cm')
+
+    def test_flip(self):
+        raw_box = Mesh(vc('/unittest/serialization/obj/test_box.obj'))
+        box = Mesh(v=raw_box.v, f=raw_box.f)
+        box.reset_normals()
+        original_v = box.v.copy()
+        original_vn = box.vn.copy()
+        original_f = box.f.copy()
+        box.flip(axis=0)
+        box.reset_normals()
+        self.assertEqual(box.f.shape, original_f.shape)
+        self.assertEqual(box.v.shape, original_v.shape)
+        for face, orig_face in zip(box.f, original_f):
+            self.assertNotEqual(list(face), list(orig_face))
```

Comment:
So the important part of the test is actually the fact that the normals are flipped. All this chunk is really doing is asserting that the faces are the thing that changed. The goal of the function being tested is actually to invert the normals, even though the name of the function is `flip_faces`, which describes the method by which the normals are inverted. 

There's a principle of testing that's being exposed here, which is that it's generally preferable to make your tests as weak as possible while still capturing the pre and post conditions of the thing tested. The idea is that the test should allow you to refactor the function. If the test captures the internal logic of the function as currently implemented, then you can't use it to refactor the function. Also, if you got that logic wrong then you're testing the wrong thing. So the perfect test actually doesn't know how the function is implemented (and ideally is actually spec'd and maybe written before the function). Tightly coupling the test and the implementation loses much of the benefit of unit testing.

So yes, the function is deterministic and we could essentially assert that the implementation is never going to change. But if I were to change these tests, I'd actually be inclined to remove the testing of faces and _only_ test that the normals are inverted, as that's the key feature of such a function. As long as the normals are being inverted, the function is meeting its post-condition, regardless of how it's actually managed.